### PR TITLE
fix(imports): Specify moduleName overrides for imports

### PR DESCRIPTION
**Pull request by campionfellin**
Fixes: https://github.com/awslabs/cdk8s/issues/69

Signed-off-by: campionfellin <campionfellin@gmail.com>

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: examples/crd/cdk8s.yaml [hunk 1]
```diff
@@ -3,4 +3,5 @@ language: typescript
 imports:
   - k8s@1.17.0
   - https://raw.githubusercontent.com/jenkinsci/kubernetes-operator/master/deploy/crds/jenkins.io_jenkins_crd.yaml
-  - mattermost_crd.yaml
+  - moduleName: mattermost
+    file: mattermost_crd.yaml
```
Comment (campionfellin) [inline examples/crd/cdk8s.yaml]: This somewhat acts as a test, but I'm wondering if we want more robust tests to cover this in `./tests/` or in `./packages/cdk8s-cli/test`?

Comment (eladb) [reply examples/crd/cdk8s.yaml]: Yes. Definitely!

Comment (campionfellin) [reply examples/crd/cdk8s.yaml]: Awesome, will do that then :)

**Comment by campionfellin** [comment]
Ok! I will rework it a bit :)

File: packages/cdk8s-cli/lib/config.ts [hunk 2]
```diff
@@ -4,11 +4,16 @@ import { Language } from './import/base';
 
 const CONFIG_FILE = 'cdk8s.yaml';
 
+export interface Cdk8sImport {
```
Comment (eladb) [inline packages/cdk8s-cli/lib/config.ts:7]: Rename to `ImportSpec`

File: packages/cdk8s-cli/lib/config.ts [hunk 3]
```diff
@@ -4,11 +4,16 @@ import { Language } from './import/base';
...
+  moduleName?: string;
```
Comment (eladb) [inline packages/cdk8s-cli/lib/config.ts:8]: Rename to `name` and `source` (it’s not always a file... could be url or k8s...)

File: packages/cdk8s-cli/lib/import/base.ts [hunk 4]
```diff
@@ -22,15 +23,15 @@ export abstract class ImportBase {
   public abstract get moduleNames(): string[];
   protected abstract async generateTypeScript(code: CodeMaker, moduleName?: string): Promise<void>;
 
-  public async import(options: ImportOptions) {
+  public async import(options: ImportOptions, source?: Cdk8sImport) {
```
Comment (eladb) [inline packages/cdk8s-cli/lib/import/base.ts:26]: The `spec` should go in the constructor instead of `moduleName` there.

File: packages/cdk8s-cli/bin/cmds/import.ts [hunk 5]
```diff
@@ -33,4 +35,32 @@ class Command implements yargs.CommandModule {
...
+  // crd=crd.yaml
+  // crd=url.com/crd.yaml
+  if (splitImport.length === 2) {
+    return {
+      moduleName: splitImport[0],
+      file: splitImport[1]
+    }
+  }
+
+  // We'll assume that no one hosts remote files with query params,
```
[the remark this answers is not in the document]
Comment (campionfellin) [reply packages/cdk8s-cli/bin/cmds/import.ts:58]: Do you know of any good way to check this? I was thinking something like:

```
const firstEquals = file.indexOf('=');

// if first equals is after http:// or https:// then ignore
if (firstEquals > file.indexOf('http://')) {
  return {
    file
  }
}

if (firstEquals > file.indexOf('https://')) {
  return {
    file
  }
}
```

But I'm wondering if that's a surefire way to know.

[the remark this answers is not in the document]
Comment (eladb) [reply packages/cdk8s-cli/bin/cmds/import.ts:58]: You are right, parsing this is unnecessary hard. Let's just use another delimiter such as `:=` to simplify?

```
foo:=crd.yaml
bar:=https://foo.bar/c.crd?q=2
```

What do you think? Any other preferred delimiter?

[the remark this answers is not in the document]
Comment (campionfellin) [reply packages/cdk8s-cli/bin/cmds/import.ts:58]: That seems fine to me! Do you think command line will complain about it?

`cdk8s import foo:=bar.yml -l typescript `

(I can check if it works, but it does look a bit off)

File: packages/cdk8s-cli/lib/import/base.ts (hunk 4 above)
Comment (campionfellin) [reply packages/cdk8s-cli/lib/import/base.ts:26]: Sorry, I'm not following this one. What do you mean here?

File: packages/cdk8s-cli/bin/cmds/import.ts [hunk 6]
```diff
@@ -35,31 +35,31 @@ class Command implements yargs.CommandModule {
   }
 }
 
-function transformImports(file: string): Cdk8sImport {
+function parseImports(file: string): ImportSpec {
```
Comment (eladb) [inline packages/cdk8s-cli/bin/cmds/import.ts:38]:
```suggestion
function parseImports(spec: string): ImportSpec {
```

File: packages/cdk8s-cli/lib/import/crd.ts [hunk 7]
```diff
@@ -85,8 +85,8 @@ export class CustomResourceDefinition {
 }
 
 export class ImportCustomResourceDefinition extends ImportBase {
-  public static async match(source: Cdk8sImport): Promise<undefined | CustomResourceApiObject[]> {
-    const { file } = source;
+  public static async match(source: ImportSpec): Promise<undefined | CustomResourceApiObject[]> {
```
Comment (eladb) [inline packages/cdk8s-cli/lib/import/crd.ts:88]:
```suggestion
  public static async match(spec: ImportSpec): Promise<undefined | CustomResourceApiObject[]> {
```

File: packages/cdk8s-cli/lib/import/dispatch.ts [hunk 8]
```diff
@@ -1,21 +1,24 @@
...
-    await importer.import(options, source);
+    await importer.import({
+      ...options,
+      moduleNamePrefix: source.moduleNamePrefix
+    });
   }
 }
 
-async function matchImporter(source: Cdk8sImport, argv: any) {
+async function matchImporter(source: ImportSpec, argv: any) {
```
Comment (eladb) [inline packages/cdk8s-cli/lib/import/dispatch.ts:21]:
```suggestion
async function matchImporter(spec: ImportSpec, argv: any) {
```

File: packages/cdk8s-cli/lib/import/k8s.ts [hunk 9]
```diff
@@ -32,8 +32,8 @@ export interface ImportKubernetesApiOptions {
 
 export class ImportKubernetesApi extends ImportBase {
 
-  public static async match(source: Cdk8sImport, argv: any): Promise<ImportKubernetesApiOptions | undefined> {
-    const { file } = source;
+  public static async match(source: ImportSpec, argv: any): Promise<ImportKubernetesApiOptions | undefined> {
```
Comment (eladb) [inline packages/cdk8s-cli/lib/import/k8s.ts:35]:
```suggestion
  public static async match(spec: ImportSpec, argv: any): Promise<ImportKubernetesApiOptions | undefined> {
```

File: packages/cdk8s-cli/lib/import/base.ts (hunk 4 above)
Comment (campionfellin) [reply packages/cdk8s-cli/lib/import/base.ts]: I think this is resolved now.. `source` is now named `importSpec`

File: packages/cdk8s-cli/bin/cmds/import.ts [hunk 10]
```diff
@@ -14,7 +14,7 @@ class Command implements yargs.CommandModule {
 
   public readonly builder = (args: yargs.Argv) => args
     .positional('SOURCE', { default: config.imports, desc: `The API source to import (supported sources: k8s, crd.yaml, https://domain/crd.yaml)`, array: true })
```
[the remark this answers is not in the document]
Comment (eladb) [reply packages/cdk8s-cli/bin/cmds/import.ts:16]: ?

File: packages/cdk8s-cli/bin/cmds/import.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (campionfellin) [reply packages/cdk8s-cli/bin/cmds/import.ts:59]: Ok, this seems to work now, but without changes to `yargs`.

File: packages/cdk8s-cli/bin/cmds/import.ts [hunk 11]
```diff
@@ -59,7 +60,7 @@ function parseImports(file: string): ImportSpec {
   // like this: url.com/crds?crd=mycrd.yaml
   console.error('If you reach here, please open an issue with cdk8s.');
```
Comment (eladb) [inline packages/cdk8s-cli/bin/cmds/import.ts:61]: Throw an exception: "unable to parse import specification. syntax is [NAME=]SOURCE

File: packages/cdk8s-cli/lib/config.ts [hunk 12]
```diff
@@ -5,8 +5,8 @@ import { Language } from './import/base';
 const CONFIG_FILE = 'cdk8s.yaml';
 
 export interface ImportSpec {
-  readonly moduleNamePrefix?: string;
-  readonly source: string;
+  moduleNamePrefix?: string;
```
Comment (eladb) [inline packages/cdk8s-cli/lib/config.ts:8]: Why did you remove `readonly`?

File: packages/cdk8s-cli/lib/import/base.ts [hunk 13]
```diff
@@ -23,16 +24,15 @@ export abstract class ImportBase {
   public abstract get moduleNames(): string[];
   protected abstract async generateTypeScript(code: CodeMaker, moduleName?: string): Promise<void>;
 
-  public async import(options: ImportOptions) {
+  public async import(options: ImportOptions, importSpec?: ImportSpec) {
```
Comment (eladb) [inline packages/cdk8s-cli/lib/import/base.ts:27]: Why not put `moduleNamePrefix` in `ImportOption`?

File: packages/cdk8s-cli/bin/cmds/import.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (campionfellin) [reply packages/cdk8s-cli/bin/cmds/import.ts:16]: Could not get this working :(

What about this syntax?

```
cdk8s import --k8s=k8s (saves it to `k8s-k8s.ts`)
cdk8s import --myCrdModule=crd.yml (saves to `myCrdModule-crd.ts`
cdk8s import k8s (saves to `k8s.ts`)
cdk8s import --abc=k8s --other=crd.yaml (saves 2 files, `abc-k8s.ts` and `other-crd.ts`)
```

File: packages/cdk8s-cli/lib/config.ts (hunk 12 above)
Comment (campionfellin) [reply packages/cdk8s-cli/lib/config.ts:8]: Oops, merging error!

File: packages/cdk8s-cli/lib/import/base.ts (hunk 13 above)
Comment (campionfellin) [reply packages/cdk8s-cli/lib/import/base.ts:27]: Ahh, I see it now. Missed it during the merge.

Comment (campionfellin) [reply packages/cdk8s-cli/lib/import/base.ts:27]: And now the comment from above makes sense :)

File: packages/cdk8s-cli/bin/cmds/import.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (eladb) [reply packages/cdk8s-cli/bin/cmds/import.ts:16]: Forget about it. Just change `SOURCE` to `SPEC` then and describe the syntax in the `description` of `SPEC`

[the remark this answers is not in the document]
Comment (eladb) [reply packages/cdk8s-cli/bin/cmds/import.ts:16]:
```suggestion
    .positional('SPEC', { default: config.imports, desc: `The import spec with the syntax [NAME=]SOURCE where NAME is an optional module name and supported SOURCEs are: k8s, crd.yaml, https://domain/crd.yaml`, array: true })
```

Something like that.

File: packages/cdk8s-cli/bin/cmds/import.ts [hunk 14]
```diff
@@ -13,24 +13,50 @@ class Command implements yargs.CommandModule {
   public readonly aliases = [ 'gen', 'import', 'generate' ];
 
   public readonly builder = (args: yargs.Argv) => args
-    .positional('SOURCE', { default: config.imports, desc: `The API source to import (supported sources: k8s, crd.yaml, https://domain/crd.yaml)`, array: true })
-    .example(`cdk8s import k8s`, `Imports Kubenretes API objects to imports/k8s.ts`)
-    .example(`cdk8s import k8s@1.13.0`, `imports a specific version of the Kubenretes API`)
-    .example(`cdk8s import jenkins.io_jenkins_crd.yaml`, 'imports constructs for the Jenkins custom resource definition from a file')
+    .positional('SPEC', { default: config.imports, desc: `import spec with the syntax [NAME=]SOURCE where NAME is an optional module name and supported SOURCEs are: k8s, crd.yaml, https://domain/crd.yaml)`, array: true })
+    .example(`cdk8s import --k8s=k8s`, `Imports Kubernetes API objects to imports/k8s.ts`)
```
Comment (campionfellin) [inline packages/cdk8s-cli/bin/cmds/import.ts:17]:
```suggestion
    .example(`cdk8s import k8s`, `Imports Kubernetes API objects to imports/k8s.ts`)
```

File: packages/cdk8s-cli/bin/cmds/import.ts [hunk 15]
```diff
@@ -13,24 +13,50 @@ class Command implements yargs.CommandModule {
   public readonly aliases = [ 'gen', 'import', 'generate' ];
 
   public readonly builder = (args: yargs.Argv) => args
-    .positional('SOURCE', { default: config.imports, desc: `The API source to import (supported sources: k8s, crd.yaml, https://domain/crd.yaml)`, array: true })
-    .example(`cdk8s import k8s`, `Imports Kubenretes API objects to imports/k8s.ts`)
-    .example(`cdk8s import k8s@1.13.0`, `imports a specific version of the Kubenretes API`)
-    .example(`cdk8s import jenkins.io_jenkins_crd.yaml`, 'imports constructs for the Jenkins custom resource definition from a file')
+    .positional('SPEC', { default: config.imports, desc: `import spec with the syntax [NAME=]SOURCE where NAME is an optional module name and supported SOURCEs are: k8s, crd.yaml, https://domain/crd.yaml)`, array: true })
```
Comment (campionfellin) [inline packages/cdk8s-cli/bin/cmds/import.ts:16]: Thinking the config files should have a list of `importSpecs` instead of `imports`. Thoughts?

**Comment by campionfellin** [comment]
Will look into the python integration test failures tomorrow!

File: packages/cdk8s-cli/bin/cmds/import.ts (hunk 15 above)
Comment (eladb) [reply packages/cdk8s-cli/bin/cmds/import.ts:16]: leave as `imports`

Comment (eladb) [inline packages/cdk8s-cli/bin/cmds/import.ts:16]:
```suggestion
    .positional('SPEC', { default: config.imports, desc: `import spec with the syntax [NAME:=]SOURCE where NAME is an optional module name and supported SOURCEs are: k8s, crd.yaml, https://domain/crd.yaml)`, array: true })
```

File: packages/cdk8s-cli/lib/config.ts [hunk 16]
```diff
@@ -4,11 +4,16 @@ import { Language } from './import/base';
...
+  readonly moduleNamePrefix?: string;
+  readonly source: string;
+}
+
 export interface Config {
   readonly app?: string;
   readonly language?: Language;
   readonly output?: string;
-  readonly imports?: string[];
+  readonly imports?: (string|ImportSpec)[];
```
Comment (eladb) [inline packages/cdk8s-cli/lib/config.ts:16]: This still needs to be strings I think

File: packages/cdk8s-cli/bin/cmds/import.ts (hunk 15 above)
Comment (campionfellin) [reply packages/cdk8s-cli/bin/cmds/import.ts]: Sounds good to me!

**Comment by eladb** [comment]
Integ tests is failing https://github.com/awslabs/cdk8s/runs/568224322

**Comment by campionfellin** [comment]
Yes, I can add a couple more tests to the `test-imports/`!

I think it'd also be useful to have some of the examples be written in Python!

File: packages/cdk8s-cli/lib/import/base.ts [hunk 17]
```diff
@@ -75,8 +76,8 @@ export abstract class ImportBase {
   
   }
 
-  private async harvestPython(targetdir: string, moduleName: string) {
-    const target = path.join(targetdir, moduleName);
+  private async harvestPython(targetdir: string, moduleName: string, moduleNamePrefix?: string) {
+    const target = path.join(targetdir, moduleNamePrefix ? moduleNamePrefix + "." + moduleName : moduleName);
```
Comment (campionfellin) [inline packages/cdk8s-cli/lib/import/base.ts:80]:
```suggestion
    const target = path.join(targetdir, moduleNamePrefix ? `${moduleNamePrefix}.${moduleName}` : moduleName);
```

File: packages/cdk8s-cli/bin/cmds/import.ts [hunk 18]
```diff
@@ -8,29 +8,55 @@ const DEFAULT_OUTDIR = 'imports';
...
+  // crd=crd.yaml
+  // crd=url.com/crd.yaml
+  if (splitImport.length === 2) {
+    return {
+      moduleNamePrefix: splitImport[0],
+      source: splitImport[1]
+    }
+  }
+
+  throw new Error(`Unable to parse import specification. Syntax is [NAME:=]SOURCE`);
```
Comment (campionfellin) [inline packages/cdk8s-cli/bin/cmds/import.ts:59]:
```suggestion
  throw new Error(`Unable to parse import specification. Syntax is [NAME:=]SPEC`);
```

**Comment by campionfellin** [comment]
As discussed offline, we'll add python support at a later time. For now, we'll throw error:

```
Name overriding of imports is not yet supported in python. Named import: <import name>
```

File: test/test-imports/python/test.sh [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+set -e
+scriptdir=$(cd $(dirname $0) && pwd)
+
+cd $(mktemp -d)
+mkdir test && cd test
+
+cp ${scriptdir}/cdk8s.yaml  .
+cp ${scriptdir}/mattermost_crd.yaml .
+
+cdk8s import mattermost:=mattermost_crd.yaml --language python || echo "Named import not supported yet."
```
Comment (campionfellin) [inline test/test-imports/python/test.sh:11]: This is working fine locally. Will try to update to get github build to work. Probably something like `2> /dev/null` should work

**Comment by campionfellin** [comment]
Integ tests were failing because I did not commit the `.tgz` files that were needed in the diffs. 

Committed them, and they fail because the headers differ (I think this is because they are zipped at different times, so the timestamp is different).

Skip the `.tgz` files in the diff.
